### PR TITLE
✨ Add random extensions.

### PR DESCRIPTION
**Pull request by fortmarek**
#### Checklist
- [x] Added tests (if applicable)

Added random extensions that can be used in tests, but also in a more general-purpose way.

There are a lot of types missing, but so far we have been able to go with just those that I have added. It should be pretty straightforward to add additional ones if such a need arises.

### CODE REVIEW

File: ACKategoriesCore/RandomExtensions/Int+Random.swift [hunk 1]
```diff
@@ -0,0 +1,24 @@
+import Foundation
+
+public extension Int {
+    /// - Parameters:
+    ///     - min: Minimal value of returned Int
+    ///     - max: Maximal value of returned Int
+    /// - Returns random Int
+    static func random(min: Int = 20, max: Int = 1000) -> Int {
```
Comment (LukasHromadnik) [inline ACKategoriesCore/RandomExtensions/Int+Random.swift:8]: ❓ what's the added value of this method when we can use `Int.random(in range: Range<Int>)`?

Comment (fortmarek) [reply ACKategoriesCore/RandomExtensions/Int+Random.swift:8]: None, forgot about this one 😛

Comment (fortmarek) [reply ACKategoriesCore/RandomExtensions/Int+Random.swift:8]: Well, actually it's nice to have some sensible defaults, so you don't have to input the range everywhere. But I'll rewrite it, so it uses your mentioned method under the hood.

File: ACKategoriesCore/RandomExtensions/Date+Random.swift [hunk 2]
```diff
@@ -0,0 +1,23 @@
+import Foundation
+
+public extension Date {
+    /// - Parameters:
+    ///     - min: Minimum `timeIntervalSince1970`
+    ///     - max: Maxium `timeIntervalSince1970`
+    /// - Returns: Random date
+    static func random(min: Int = 0, max: Int = 2000000000) -> Date {
```
Comment (janmisar) [inline ACKategoriesCore/RandomExtensions/Date+Random.swift:8]: Why `2000000000`? It's just a detail but maybe we could use `Int.max`?

Comment (fortmarek) [reply ACKategoriesCore/RandomExtensions/Date+Random.swift:8]: Yeah, I agree that would make sense 👍

File: ACKategoriesCore/RandomExtensions/Int+Random.swift [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+        return Int(arc4random_uniform(UInt32(max - min + 1))) + min
+    }
+}
+
+public extension Collection where Element == Int {
+    /// - Parameters:
+    ///     - min: Minimal number of returned elements
+    ///     - max: Maximal number of returned elements
+    /// - Returns: Array of random Ints
+    static func random(min: Int = 20, max: Int = 100) -> [Element] {
```
Comment (LukasHromadnik) [inline ACKategoriesCore/RandomExtensions/Int+Random.swift:19]: 💡 we can create a protocol to connect all types with `.random()` method and then use only one implementation for a random collection. What do you think?

Comment (fortmarek) [reply ACKategoriesCore/RandomExtensions/Int+Random.swift:19]: I don't think that will work since the methods are static ... ?

File: ACKategoriesCore/RandomExtensions/String+Random.swift [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+    public static func random() -> String {
+        String.random(minLength: 5)
+    }
+    
+    /// - Parameters:
+    ///     - minLength: Minimum length of a returned String
+    ///     - maxLength: Maxium length of a returned String
+    ///     - allowedLetters: Letters that will be used in the randomization
+    /// - Returns: Randomized String
+    public static func random(minLength: Int = 5,
```
Comment (olejnjak) [inline ACKategoriesCore/RandomExtensions/String+Random.swift:13]: I am just wondering, if we have `Collection`s comforming to `Randomizable`, would it be possible to take advantage of `String`s being collections?

File: ACKategoriesCore/RandomExtensions/String+Random.swift [hunk 5]
```diff
@@ -0,0 +1,24 @@
...
+    }
+
+    /// - Parameters:
+    ///     - minLength: Minimum length of a returned String
+    ///     - maxLength: Maxium length of a returned String
+    ///     - allowedLetters: Letters that will be used in the randomization
+    /// - Returns: Randomized String
+    public static func random(minLength: Int = 5,
+                              maxLength: Int = 10,
+                              allowedLetters: String = "abcdefghijklmnopqrstuvwxyzěščřžýáíéABCDEFGHIJKLMNOPQRSTUVWXYZĚŠČŘŽÝÁÍÉ0123456789") -> String {
```
Comment (michalsrutek) [inline ACKategoriesCore/RandomExtensions/String+Random.swift:15]: It says `allowedLetters`, but I can also see digits in there. Maybe `allowedCharacters` would be better?

File: ACKategoriesCore/RandomExtensions/Randomizable.swift [hunk 6]
```diff
@@ -0,0 +1,20 @@
...
+    static func random() -> Self
+}
+
+public extension Collection where Element: Randomizable {
+    /// - Parameters:
+    ///     - min: Minimal number of returned elements
+    ///     - max: Maximal number of returned elements
+    /// - Returns: Array of random elements
+    static func random(min: Int = 20, max: Int = 100) -> [Element] {
+        assert(min <= max, "Min must be smaller than max")
```
Comment (michalsrutek) [inline ACKategoriesCore/RandomExtensions/Randomizable.swift:16]: Assertion and the message say two different things. You could 
- say something like "Min must not be greater than max", 
- say something like "Min must be smaller than or equal to max", 
- require that min is strictly lower than max

File: ACKategoriesCore/RandomExtensions/Randomizable.swift [hunk 7]
```diff
@@ -0,0 +1,20 @@
...
+        let count = (Int.random() % max) + min
+        return (0..<count).map { _ in .random() }
+    }
+}
```
Comment (michalsrutek) [inline ACKategoriesCore/RandomExtensions/Randomizable.swift:20]: This code is flawed for a case when `min = max`. In that case, one can get a wrong number of results (up to `2 * max - 1`).
I'd also suggest you create a test for this (edge) case.

File: ACKategoriesCoreTests/Random/StringRandomTests.swift [hunk 8]
```diff
@@ -0,0 +1,26 @@
...
+        DispatchQueue.concurrentPerform(iterations: 50) { _ in
+            let randomString = String.random(allowedLetters: "ad")
+            XCTAssertTrue(randomString.allSatisfy("ad".contains))
+        }
+    }
+    
+    func testLengthOfStringIsSmallerThanGivenValue() {
+        DispatchQueue.concurrentPerform(iterations: 50) { _ in
+            let randomString = String.random(minLength: 0, maxLength: 2)
+            print(randomString.count)
```
Comment (michalsrutek) [inline ACKategoriesCoreTests/Random/StringRandomTests.swift:15]: Probably for testing purposes. Can be deleted?

File: ACKategoriesCore/RandomExtensions/Date+Random.swift [hunk 9]
```diff
@@ -0,0 +1,15 @@
...
+        Date.random(min: 0)
+    }
+
+    /// - Parameters:
+    ///     - min: Minimum `timeIntervalSince1970`
+    ///     - max: Maxium `timeIntervalSince1970`
+    /// - Returns: Random date
+    public static func random(min: Int = 0, max: Int = Int.max) -> Date {
+        return Date(timeIntervalSince1970: TimeInterval(Int.random(in: min...max)))
+    }
```
Comment (michalsrutek) [inline ACKategoriesCore/RandomExtensions/Date+Random.swift:14]: I'd suggest you go for consistency here and remove the `return` keyword from the second function.

File: ACKategoriesCore/RandomExtensions/String+Random.swift [hunk 10]
```diff
@@ -0,0 +1,24 @@
...
+        let length = Int.random(in: minLength...maxLength)
+        return (0..<length).map { _ in
+            let position = Int.random(in: 0..<allowedLetters.count)
+            let index = allowedLetters.index(allowedLetters.startIndex, offsetBy: position)
+            return String(allowedLetters[index])
+        }
+        .joined()
+    }
```
Comment (michalsrutek) [inline ACKategoriesCore/RandomExtensions/String+Random.swift:23]: I'd add some sanity checks here, especially to ensure that
- letters (characters) are not empty if we want to generate a non-empty String
- lengths are not negative
- `minLength <= maxLength`

File: ACKategoriesCore/RandomExtensions/String+Random.swift (hunk 5 above)
Comment (fortmarek) [reply ACKategoriesCore/RandomExtensions/String+Random.swift]: Sure, good idea 👍

File: ACKategoriesCore/RandomExtensions/Randomizable.swift (hunk 6 above)
Comment (fortmarek) [reply ACKategoriesCore/RandomExtensions/Randomizable.swift]: Deleted the assert altogether, as all the edge cases will be caught in `Int.random(in: min...max)`

File: ACKategoriesCore/RandomExtensions/Randomizable.swift (hunk 7 above)
Comment (fortmarek) [reply ACKategoriesCore/RandomExtensions/Randomizable.swift]: Used `Int.random(in: ...)`, which should fix it, good catch 👍 
Negative values are not allowed anymore (it will fail in `(0..<count)`)

File: ACKategoriesCoreTests/Random/StringRandomTests.swift (hunk 8 above)
Comment (fortmarek) [reply ACKategoriesCoreTests/Random/StringRandomTests.swift]: Yep 🙈

File: ACKategoriesCore/RandomExtensions/Date+Random.swift (hunk 9 above)
Comment (fortmarek) [reply ACKategoriesCore/RandomExtensions/Date+Random.swift]: why not 🆗

File: ACKategoriesCore/RandomExtensions/String+Random.swift (hunk 10 above)
Comment (fortmarek) [reply ACKategoriesCore/RandomExtensions/String+Random.swift]: All of the issues will surface in creating the ranges, so it's better not to add additional logic on our side when it's not necessary.
